src/Microsoft.ML.Data/Scorers/FeatureContributionCalculationTransform.cs

```diff
@@ -4,35 +4,55 @@
 
 using System;
```

Comment:
I feel like the majority of required changes are in this file. This we discussed offline, but wanted to codify in 

While we need to have the scorer still for command line usage, we have an advantage that the command line tool does not have. In the command line tool we *had* to phrase it as a scorer, because of the fixed pattern in which train-score-evaluate *had* to happen on the command line. However in API usage, all we really need is to make the feature contribution scores just another transformer, without bundling it with the scorers.

Because the command line can only invoke one scorer at a time, the original code had to solve what was actually a fairly complex task: bundle the task specific scorer *and* the feature contributions (this, again, is because there could be only one scorer).

In the case of the API we have no such restrictions, and indeed the workflow is far simpler: we have already applied the scorer in the pipeline at the point where we are in a position to instantiate the `ITransformer`, so we already have a scorer. And what could be more natural than just applying this transformer on top of this original model? So not only is this going to be simpler to implement (from the API perspective), it will result in a more natural outcome, since otherwise in order to avoid having the scorer applied *twice* (which was the consequence of the design here), you'd have to perform a fairly elaborate set of rituals.

We still probably need this scorer (in its original form), so that it can serve the command line scenario, but we should have the simpler interface for the API.

Reply:
Thank you for writing it up Tom! I agree with the above.

---
In reply to: [242659979](https://github.com/dotnet/machinelearning/pull/1847#discussion_r242659979) [](ancestors = 242659979)